tremor-script/src/ast/deploy/raw.rs

```diff
@@ -0,0 +1,785 @@
+                }
+                DeployStmtRaw::CreateStmt(stmt) => {
+                    let stmt: CreateStmt<'script> = stmt.up(helper)?;
+                    helper
+                        .instances
+                        .insert(stmt.fqsn(&stmt.module), stmt.clone());
+                }
+            }
+        }
+        helper.module.pop();
+        Ok(())
+    }
+}
+
+/// we're forced to make this pub because of lalrpop
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub struct PipelineDeclRaw<'script> {
+    pub(crate) start: Location,
+    pub(crate) end: Location,
+    pub(crate) id: String,
+    pub(crate) args: WithArgsRaw<'script>,
+    pub(crate) params: Option<DeployWithExprsRaw<'script>>,
+    pub(crate) query: QueryRaw<'script>,
+}
+
+fn arg_spec_resolver<'script, 'registry>(
+    args_spec: WithArgsRaw<'script>,
+    params_spec: &Option<DeployWithExprsRaw<'script>>,
+    helper: &mut Helper<'script, 'registry>,
+) -> Result<(Vec<Cow<'script, str>>, Value<'script>)> {
+    let mut args: HashMap<Cow<str>, Value<'script>> = HashMap::new();
+    let mut spec = vec![];
+
+    let upper_args = helper.consts.args.clone();
+
+    // Process explicitly declared required/optional arguments
+    for arg_spec in args_spec {
+        match arg_spec {
+            ArgumentDeclRaw::Required(name) => {
+                let moo = name.to_string();
+                let moo: Cow<str> = Cow::owned(moo);
+                spec.push(moo.clone());
+            }
+            ArgumentDeclRaw::Optional(name, value) => {
+                let moo = name.to_string();
+                let moo: Cow<str> = Cow::owned(moo);
+                spec.push(moo.clone());
+                let value = value.up(helper)?;
+                let value = stateless_run_expr!(helper, &value);
+                args.insert(moo, value);
+            }
+        }
+    }
+    // We inject the args here so that our `with` clause can override based on
+    // our specificational arguments and their defaults
+    helper.consts.args = Value::Object(Box::new(args.clone()));
+
+    // Process defaulted arguments via `with` expressions
+    if let Some(params_spec) = up_maybe_params(params_spec, helper)? {
+        for (name, value) in params_spec {
+            let moo = Cow::owned(name);
+            if !spec.contains(&moo) {
+                // FIXME TODO consider duplicate argument specification - ok or ko?
+                spec.push(moo.clone());
+            }
+            let up_value = value;
+            args.insert(moo, up_value.clone_static());
+        }
+        // We inject the args here so that our `script` clause can override based on
+        // our `with` specificational argument overrides and their defaults
+        helper.consts.args = Value::Object(Box::new(args.clone()));
+    }
+
+    // NOTE We don't check for `extra args` here - this is deferred to the resolution in `arg_resolver`
+    helper.consts.args = upper_args.clone();
```

Comment:
I'm a bit confused about what happens where, we set args, then set it again, and then set it a third time, apparently without reading it in between. I think I'm missing some context here?

Reply:
This was harder to get behaving correctly than it looks - if the args aren't updated again we will see the wrong arguments at various points in subsequent parsing. This is one attempt to correct that - the other was making helper cloneable. I suspect there's a better way to achieve this but there are 1 or 2 other runtime side-effects that still need to be resolved ( helper again ) - once those are patched we can refactor.

Reply:
Outdated, and the args/with refactoring replaces this. Closing.